client/src/pages/private/participant-details/ParticipantDetailsView.js

```diff
@@ -0,0 +1,385 @@
+// Participant Details Page
+// Dependency
+import React, { useEffect, useState, useMemo } from 'react';
+import { useParams, useHistory } from 'react-router-dom';
+import _orderBy from 'lodash/orderBy';
+
+import { Box, Card, Grid, Link, Typography } from '@material-ui/core';
+import { makeStyles } from '@material-ui/core/styles';
+import EditIcon from '@material-ui/icons/Edit';
+
+// Libs
+import { getLinkLabel, keyLabelMap, displayData, fetchData } from './constants';
+import { AssignCohortDialog, EditParticipantDialog, EditRosFieldDialog } from './modals';
+import { useToast } from '../../../hooks';
+import { AuthContext } from '../../../providers';
+import { Page, CheckPermissions, Alert, Button } from '../../../components/generic';
+import {
+  ToastStatus,
+  Routes,
+  EditRosDateSchema,
+  EditRosSiteSchema,
+  EditRosStartDateSchema,
+  mohEditType,
+  MAX_LABEL_LENGTH,
+} from '../../../constants';
+import {
+  assignParticipantWithCohort,
+  updateParticipant,
+  onRosDateUpdate,
+  onRosSiteUpdate,
+  onRosStartDateUpdate,
+} from '../../../services';
+import { addEllipsisMask } from '../../../utils';
+
+// Sub component
+import { PSICohortView } from '../../../components/participant-details';
+
+const useStyles = makeStyles((theme) => ({
+  root: {
+    marginBottom: theme.spacing(2),
+    padding: theme.spacing(4),
+  },
+  gridSection: {
+    paddingTop: theme.spacing(4),
+    paddingBottom: theme.spacing(4),
+  },
+}));
+
+export default () => {
+  // History
+  const history = useHistory();
+  // State
+  const [error, setError] = useState(null);
+  const [participant, setParticipant] = useState(null);
+  const [actualParticipant, setActualParticipant] = useState(null);
+  const [showEditModal, setShowEditModal] = useState(false);
+  const [psiList, setPSIList] = useState([]);
+  const [disableAssign, setDisableAssign] = useState(false);
+  const [selectedCohort, setSelectedCohort] = useState(null);
+  const [editFormField, setEditFormField] = useState(null);
+  const [allSites, setAllSites] = useState([]);
+  // Hook: Toast
+  const { openToast } = useToast();
+  // Auth context
+  const { auth } = AuthContext.useAuth();
+  // Memo roles
+  const roles = useMemo(() => auth.user?.roles || [], [auth.user?.roles]);
+  // Style classes
+  const classes = useStyles();
+  // Get param
+  const { id, page, pageId } = useParams();
+  // Breadcrumb name
+  const linkName = getLinkLabel(page);
+  // Edit Button flag
+  const enableEdit = roles.some((role) => ['ministry_of_health', 'superuser'].includes(role));
+  const isMoH = roles.includes('ministry_of_health');
+
+  const rosKeyMap = {
```

Comment:
What you've done with the edit form is pretty cool, but I wonder if it's a bit too much complexity in the name of reusing the form. From the perspective of someone wanting to modify the "Edit site name form" for example, they need to understand quite a bit of logic and risk affecting other forms with their changes.

Would defining each of these forms individually be better?

Reply:
Makes sense 👍 I think it's best to have separate forms for each field